remove js.Wrapper and add time.Time convenience function

nitram509 opened the pull request:
Hi,

this is a great lib.  While tinkering around, I found two issues, I would like to share/improve.
This PR covers two aspects. 

* with Go 1.18, the js.Wrapper is no more available, so i removed it. See https://tip.golang.org/doc/go1.18
* for type time.Time, a "new Date" object in JS is created.

Any feedback is welcome.

nitram509 committed:
remove js.Wrapper, because removed with Go 1.18 onwards,
see Go 1.18 release notes: https://tip.golang.org/doc/go1.18

nitram509 committed:
add convenience method for time.Time conversion

nitram509 committed:
add 1.18 and 1.19 in the test matrix

nitram509 committed:
fix time.Time detection, because it's a struct

nitram509 committed:
add test for time.Time conversion

norunners commented on go.mod:
```diff
@@ -1,3 +1,5 @@
 
 go 1.16
+
+require github.com/corbym/gocrest v1.0.5 // indirect
```
Remove this test dependency.

nitram509 replied:
It's not possible to remove this one - it would break the provided test below.
I guess we're talking about introducing a new dependency here?
I'm sorry I forgot to mention it in the main description.
May I propose adding this Gocrest library, as it significantly improves the readability of tests.
Here's an example
```go
then.AssertThat(testing, "hi", is.EqualTo("bye").Reason("we are going"))
```
that's way more expressive, compare to the vanilla Golang style of writing ```if ... then t.fail()...```

What do you think?
Would it be ok to leave the test as is?

norunners replied:
This library is simple enough to have zero dependencies. I'll look into updating your tests if you like.

nitram509 replied:
Hi, thank you for answering.

Well, from engineer to engineer, I can't agree on "This library is simple enough to have zero dependencies."
Since this is a "test only" dependency and test-support libraries like this have the tendency to be stable over a long period of time. The major benefit I see is improved expressiveness when reading tests.
Hence, I consider the value of following the clean code principle is the stronger goal,
in relation to zero dependencies (as in all dependencies).
(just my 2cents)

That said, this is your lib, and you have the last word about the goals and the non-goals of this lib.

If you wish, I can rework towards "vanilla go" and remove the dependency as well, or you do this ... as you wish.

Happy to chat further, if you want :)

norunners commented on value.go:
```diff
@@ -37,6 +38,10 @@ func ValueOf(i interface{}) Value {
 
 // valueOf recursively returns a new value.
 func valueOf(v reflect.Value) js.Value {
+	if t, ok := v.Interface().(time.Time); ok {
```
Can this be pushed down into the `valueOfStruct` logic since the kind is a struct?

nitram509 replied:
fine for me ... will do later.

norunners commented:
Thank you for the contribution.

nitram509 committed:
refactor Time object transformation into valueOfStruct(), as requested by feedback

nitram509 committed:
fix go.mod + go.sum via 'go mod tidy'

nitram509 committed:
just empty line to trigger Github Action

nitram509 committed:
upgrade actions/setup-go@v3 and actions/checkout@v3, because Github warnings:
"Node.js 12 actions are deprecated. For more information see: https://github.blog/changelog/2022-09-22-github-actions-all-actions-will-begin-running-on-node16-instead-of-node12/. Please update the following actions to use Node.js 16: actions/setup-go@v2, actions/checkout@v2"
remove former empty line to trigger Github Action

nitram509 commented:
Hi @norunners 
I did refactor the code into the ```valueOfStruct()``` function, as you requested.

Additionally, I did fix the Github workflow, because of the warnings (see the screenshot attached).

May I ask, how you think about the test library 'gocrest' used? (see my previous comment)

<img width="1450" alt="Screenshot 2022-11-19 at 14 21 57" src="https://user-images.githubusercontent.com/1189394/202853248-8fd6c566-6969-44cf-b55e-00faa760016e.png">

jasikpark commented on value_test.go:
```diff
@@ -0,0 +1,22 @@
+
+	jsVal := ValueOf(now)
+
+	// cut of the timezone information, because seconds precision is sufficient for this test
```
```suggestion
	// cut off the timezone information, because second's precision is sufficient for this test
```

nitram509 committed:
Update value_test.go

Co-authored-by: Caleb Jasik <calebjasik@jasik.xyz>

jasikpark commented:
would it be possible to split this PR into multiple? one fixing go 1.18 support && one adding the testing library w/ nice helpers?

nitram509 commented:
@jasikpark @norunners I did my best in splitting this PR into three separate ones #18 #19 and #20.
So I close this one.
